appinventor/blocklyeditor/src/blocks/components.js

```diff
@@ -61,21 +68,35 @@ Blockly.Blocks.component_event = {
   domToMutation : function(xmlElement) {
 
     this.typeName = xmlElement.getAttribute('component_type');
-    this.instanceName = xmlElement.getAttribute('instance_name');//instance name not needed
     this.eventName = xmlElement.getAttribute('event_name');
     var horizParams = xmlElement.getAttribute('vertical_parameters') !== "true";
+    var isGenericString = xmlElement.getAttribute('is_generic');
+    this.isGeneric = (isGenericString == "true" ? true : false);
+    if(!this.isGeneric) {
+      this.instanceName = xmlElement.getAttribute('instance_name');//instance name not needed
+    }
 
      // Orient parameters horizontally by default
 
     this.setColour(Blockly.ComponentBlock.COLOUR_EVENT);
 
     this.componentDropDown = Blockly.ComponentBlock.createComponentDropDown(this);
-    this.componentDropDown.setValue(this.instanceName);
 
-    this.appendDummyInput('WHENTITLE').appendField(Blockly.Msg.LANG_COMPONENT_BLOCK_TITLE_WHEN)
+    if(!this.isGeneric) {
+      this.appendDummyInput('WHENTITLE')
+        .appendField(Blockly.Msg.LANG_COMPONENT_BLOCK_TITLE_WHEN)
         .appendField(this.componentDropDown, "COMPONENT_SELECTOR")
         .appendField('.' + window.parent.BlocklyPanel_getLocalizedEventName(this.getEventTypeObject().name));
-    this.componentDropDown.setValue(this.instanceName);
+      //for non-generic blocks, set the value of the component drop down
+      this.componentDropDown.setValue(this.instanceName);
+    } else {
+      this.appendDummyInput('WHENTITLE')
+        .appendField(Blockly.Msg.LANG_COMPONENT_BLOCK_TITLE_WHEN + this.typeName + '.' + window.parent.BlocklyPanel_getLocalizedEventName(this.getEventTypeObject().name));
+      var compInput = this.appendValueInput("COMPONENT")
+        .setCheck(this.typeName).appendField(Blockly.Msg.LANG_COMPONENT_BLOCK_GENERIC_METHOD_TITLE_FOR_COMPONENT)
```

Comment:
I was not sure if I should have created a new message "Blockly.Msg.LANG_COMPONENT_BLOCK_GENERIC_EVENT_TITLE_FOR_COMPONENT" (changing "METHOD" -> "EVENT"), but I was not sure if it was necessary or really exactly how to do it.